File: questions/53.MaximumSubarray.js
/**
 * maxSubArray - Greedy Algorithm
 *
 * @param {number[]} nums
 * @return {number}
 */
const maxSubArray = function (nums) {
  let max = nums[0]
  let currentSum = 0

  for (let i = 0; i < nums.length; i++) {
    if (currentSum < 0) {
      currentSum = 0
    }

    currentSum += nums[i]
    max = Math.max(max, currentSum)
  }

  return max
}

console.log(maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))

File: leetcode/53.MaximumSubarray.go
package leetcode

import "math"

func maxSubArray(nums []int) int {
	max := nums[0]
	current := 0

	for _, num := range nums {
		if current < 0 {
			current = 0
		}

		current += num
		max = int(math.Max(float64(max), float64(current)))
	}

	return max
}
